torchao/quantization/quant_api.py

```diff
@@ -507,39 +520,36 @@ def quantize_(
         m = nn.Sequential(nn.Linear(32, 1024), nn.Linear(1024, 32))
         quantize_(m, int4_weight_only(group_size=32))
 
-        # 2. write your own new apply_tensor_subclass
-        # You can also add your own apply_tensor_subclass by manually calling tensor subclass constructor
-        # on weight
-
-        from torchao.dtypes import to_affine_quantized_intx
-
-        # weight only uint4 asymmetric groupwise quantization
-        groupsize = 32
-        apply_weight_quant = lambda x: to_affine_quantized_intx(
-          x, "asymmetric", (1, groupsize), torch.int32, 0, 15, 1e-6,
-          zero_point_dtype=torch.bfloat16, preserve_zero=False, zero_point_domain="float")
-
-        def apply_weight_quant_to_linear(linear):
-            linear.weight = torch.nn.Parameter(apply_weight_quant(linear.weight), requires_grad=False)
-            return linear
-
-        # apply to modules under block0 submodule
-        def filter_fn(module: nn.Module, fqn: str) -> bool:
-            return isinstance(module, nn.Linear)
-
-        m = nn.Sequential(nn.Linear(32, 1024), nn.Linear(1024, 32))
-        quantize_(m, apply_weight_quant_to_linear, filter_fn)
-
     """
     if set_inductor_config:
         torchao.quantization.utils.recommended_inductor_config_setter()
 
-    _replace_with_custom_fn_if_matches_filter(
-        model,
-        apply_tensor_subclass,
-        _is_linear if filter_fn is None else filter_fn,
-        device=device,
-    )
+    if isinstance(config, AOBaseConfig):
+        handler = _QUANTIZE_CONFIG_HANDLER[type(config)]
+        # for each linear in the model, apply the transform if filtering passes
+        _replace_with_custom_fn_if_matches_filter(
+            model,
+            handler,
+            _is_linear if filter_fn is None else filter_fn,
+            device=device,
+            extra_args=(config,),
```

Comment:
alternatively we can pass in a lambda, then we don't need to add `extra_args` or pass in `config`:
```
replace_fn = lambda mod: handler(mod, config)
```
seems simpler

Reply:
I'm really not a fan of passing callables around, it's easy when the callable is simple but easy for future people to tack ugly stuff on and increase complexity.  Non-callable args make it harder to make the code ugly in the future.

Reply:
oh sorry, I meant pass in `replace_fn` instead of `handler`, like:
```
replace_fn = lambda mod: handler(mod, config)
_replace_with_custom_fn_if_matches_filter(
            model,
            replace_fn,
            _is_linear if filter_fn is None else filter_fn,
            device=device,
)
```
either way you're passing a callable

Reply:
hmm, still not a fan of `replace_fn = lambda mod: handler(mod, config)`.  This changes `replace_fn` from a stateless callable to a stateful callable, where the state is hard to inspect.  It's less LOC but harder to debug IMO.